pczt/src/orchard.rs

```diff
@@ -0,0 +1,548 @@
+    pub(crate) proprietary: BTreeMap<String, Vec<u8>>,
+}
+
+/// Information about an Orchard output within a transaction.
+#[serde_as]
+#[derive(Clone, Debug, Serialize, Deserialize, Getters)]
+pub struct Output {
+    //
+    // Output-specific Action effecting data.
+    //
+    // These are required fields that are part of the final transaction, and are filled in
+    // by the Constructor when adding an output.
+    //
+    pub(crate) cmx: [u8; 32],
+    pub(crate) ephemeral_key: [u8; 32],
+    /// The encrypted note plaintext for the output.
+    ///
+    /// Encoded as a `Vec<u8>` because its length depends on the transaction version.
+    ///
+    /// Once we have memo bundles, we will be able to set memos independently of Outputs.
+    /// For now, the Constructor sets both at the same time.
+    pub(crate) enc_ciphertext: Vec<u8>,
+    /// The encrypted note plaintext for the output.
+    ///
+    /// Encoded as a `Vec<u8>` because its length depends on the transaction version.
+    pub(crate) out_ciphertext: Vec<u8>,
+
+    /// The [raw encoding] of the Orchard payment address that will receive the output.
+    ///
+    /// - This is set by the Constructor.
+    /// - This is required by the Prover.
+    ///
+    /// [raw encoding]: https://zips.z.cash/protocol/protocol.pdf#orchardpaymentaddrencoding
+    #[serde_as(as = "Option<[_; 43]>")]
+    #[getset(get = "pub")]
+    pub(crate) recipient: Option<[u8; 43]>,
+
+    /// The value of the output.
+    ///
+    /// This may be used by Signers to verify that the value matches `cv`, and to confirm
+    /// the values and change involved in the transaction.
+    ///
+    /// This exposes the value to all participants. For Signers who don't need this
+    /// information, we can drop the values and compress the rcvs into the bsk global.
+    #[getset(get = "pub")]
+    pub(crate) value: Option<u64>,
+
+    /// The seed randomness for the output.
+    ///
+    /// - This is set by the Constructor.
+    /// - This is required by the Prover, instead of disclosing `shared_secret` to them.
+    #[getset(get = "pub")]
+    pub(crate) rseed: Option<[u8; 32]>,
+
+    /// The `ock` value used to encrypt `out_ciphertext`.
+    ///
+    /// This enables Signers to verify that `out_ciphertext` is correctly encrypted.
+    ///
+    /// This may be `None` if the Constructor added the output using an OVK policy of
+    /// "None", to make the output unrecoverable from the chain by the sender.
+    pub(crate) ock: Option<[u8; 32]>,
+
+    /// The ZIP 32 derivation path at which the spending key can be found for the output.
+    pub(crate) zip32_derivation: Option<Zip32Derivation>,
+
+    /// Proprietary fields related to the note being created.
+    #[getset(get = "pub")]
+    pub(crate) proprietary: BTreeMap<String, Vec<u8>>,
+}
+
+impl Bundle {
+    /// Merges this bundle with another.
+    ///
+    /// Returns `None` if the bundles have conflicting data.
+    pub(crate) fn merge(
```

Comment:
This should be a symmetric method `merge(left: Self, right: Self, ...)`; otherwise the asymmetry in `bsk` merging means you have to try the merge in both orders.

Reply:
nonblocking